### PR TITLE
HIVE-25492: Major query-based compaction is skipped if partition is empty

### PR DESCRIPTION
**Pull request by asinkovits**
### What changes were proposed in this pull request?

Do not remove empty base directories in the end of a major query based compaction.
Change how the cleaner collects base/delta/delete delta directories, which are checked if they are eligible for cleaning. 

### Why are the changes needed?

It fixes two issues.
First if the result of a major query based compaction is empty, the output (empty base directory) was deleted, so you ended up with a NOOP.
Second, the cleaner did not take into account empty directories. (base/delta/delete delta) So it did not removed old files, even when after the previous fix, the empty base was the latest. It also did not removed empty delta/delete delta directories.

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Manual + unit tests

### CODE REVIEW

File: itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/txn/compactor/TestCrudCompactorOnTez.java [hunk 1]
```diff
@@ -28,9 +28,7 @@
 
 import com.google.common.collect.Lists;
 import org.apache.commons.lang3.StringUtils;
-import org.apache.hadoop.fs.FileSystem;
-import org.apache.hadoop.fs.Path;
-import org.apache.hadoop.fs.PathFilter;
+import org.apache.hadoop.fs.*;
```
Comment (deniskuzZ) [inline itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/txn/compactor/TestCrudCompactorOnTez.java:31]: wildcard imports

File: ql/src/java/org/apache/hadoop/hive/ql/io/AcidUtils.java [hunk 2]
```diff
@@ -28,16 +28,7 @@
...
-import java.util.Arrays;
-import java.util.Collections;
-import java.util.Comparator;
-import java.util.HashMap;
-import java.util.HashSet;
-import java.util.List;
-import java.util.Map;
-import java.util.Properties;
-import java.util.Set;
+import java.util.*;
```
Comment (deniskuzZ) [inline ql/src/java/org/apache/hadoop/hive/ql/io/AcidUtils.java:31]: wildcard imports

File: ql/src/java/org/apache/hadoop/hive/ql/io/AcidUtils.java [hunk 3]
```diff
@@ -1480,6 +1471,40 @@ private static ValidTxnList getValidTxnList(Configuration conf) {
...
+    Map<Path, HdfsDirSnapshot> dirToSnapshots = new HashMap<>();
+    // depth first search
+    Deque<RemoteIterator<FileStatus>> stack = new ArrayDeque<>();
+    stack.push(fs.listStatusIterator(path));
+    while (!stack.isEmpty()) {
+      RemoteIterator<FileStatus> itr = stack.pop();
+      while (itr.hasNext()) {
+        FileStatus fStatus = itr.next();
+        Path fPath = fStatus.getPath();
+        if (acidHiddenFileFilter.accept(fPath)) {
```
Comment (deniskuzZ) [inline ql/src/java/org/apache/hadoop/hive/ql/io/AcidUtils.java:1494]: Should we skip temp dirs as well - acidTempDirFilter?

File: ql/src/java/org/apache/hadoop/hive/ql/io/AcidUtils.java [hunk 4]
```diff
@@ -1480,6 +1471,40 @@ private static ValidTxnList getValidTxnList(Configuration conf) {
...
+          if (deltaFileFilter.accept(fPath) ||
+                  baseFileFilter.accept(fPath) ||
+                  deleteEventDeltaDirFilter.accept(fPath)) {
+            dirToSnapshots.put(fPath, new HdfsDirSnapshotImpl(fPath));
+          } else {
```
Comment (deniskuzZ) [inline ql/src/java/org/apache/hadoop/hive/ql/io/AcidUtils.java:1499]: we could check if fStatus.isDirectory() before listing again

File: ql/src/java/org/apache/hadoop/hive/ql/io/AcidUtils.java [hunk 5]
```diff
@@ -1480,6 +1471,40 @@ private static ValidTxnList getValidTxnList(Configuration conf) {
...
+    // depth first search
+    Deque<RemoteIterator<FileStatus>> stack = new ArrayDeque<>();
+    stack.push(fs.listStatusIterator(path));
+    while (!stack.isEmpty()) {
+      RemoteIterator<FileStatus> itr = stack.pop();
+      while (itr.hasNext()) {
+        FileStatus fStatus = itr.next();
+        Path fPath = fStatus.getPath();
+        if (acidHiddenFileFilter.accept(fPath)) {
+          if (deltaFileFilter.accept(fPath) ||
```
Comment (deniskuzZ) [inline ql/src/java/org/apache/hadoop/hive/ql/io/AcidUtils.java:1495]: please put some order: base/delta/deleteDelta

**Review by deniskuzZ — COMMENTED** [review]
LGTM, minor comments

File: itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/txn/compactor/TestCrudCompactorOnTez.java (hunk 1 above)
Comment (asinkovits) [reply itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/txn/compactor/TestCrudCompactorOnTez.java]: Fixed.

File: ql/src/java/org/apache/hadoop/hive/ql/io/AcidUtils.java (hunk 2 above)
Comment (asinkovits) [reply ql/src/java/org/apache/hadoop/hive/ql/io/AcidUtils.java]: Fied.

File: ql/src/java/org/apache/hadoop/hive/ql/io/AcidUtils.java (hunk 3 above)
Comment (asinkovits) [reply ql/src/java/org/apache/hadoop/hive/ql/io/AcidUtils.java]: Added the filter.

File: ql/src/java/org/apache/hadoop/hive/ql/io/AcidUtils.java (hunk 4 above)
Comment (asinkovits) [reply ql/src/java/org/apache/hadoop/hive/ql/io/AcidUtils.java:1510]: Yeah, I've reiterated this part as there were issues around nonacid-acid conversion.

File: ql/src/java/org/apache/hadoop/hive/ql/io/AcidUtils.java (hunk 5 above)
Comment (asinkovits) [reply ql/src/java/org/apache/hadoop/hive/ql/io/AcidUtils.java]: I've switched the order, although I think the other way around was more optimal solution, as its more likely that an item is a delta than a base.

File: ql/src/java/org/apache/hadoop/hive/ql/io/AcidUtils.java [hunk 6]
```diff
@@ -1480,6 +1482,57 @@ private static ValidTxnList getValidTxnList(Configuration conf) {
...
+    Map<Path, HdfsDirSnapshot> dirToSnapshots = new HashMap<>();
+    // depth first search
+    Deque<RemoteIterator<FileStatus>> stack = new ArrayDeque<>();
+    stack.push(fs.listStatusIterator(path));
+    while (!stack.isEmpty()) {
+      RemoteIterator<FileStatus> itr = stack.pop();
+      while (itr.hasNext()) {
+        FileStatus fStatus = itr.next();
+        Path fPath = fStatus.getPath();
+        if (acidHiddenFileFilter.accept(fPath) && acidTempDirFilter.accept(fPath)) {
```
Comment (deniskuzZ) [inline ql/src/java/org/apache/hadoop/hive/ql/io/AcidUtils.java:1505]: we could use hiddenFileFilter as we don't need to include METADATA_FILE & ACID_FORMAT

**Review by deniskuzZ — APPROVED** [review]

**Comment by deniskuzZ** [comment]
pending tests

**Pull request merged**